### PR TITLE
feat!: replace the influx tracer with a local one

### PR DESCRIPTION
**Pull request by evan-forbes**
manual of backport of #1279

### CODE REVIEW

**Comment by evan-forbes** [comment]
wat
Screenshot from 2024-04-06 11-25-30

File: pkg/trace/fileserver.go [hunk 1]
```diff
@@ -16,10 +16,10 @@ import (
...
 	"time"
 
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/aws/credentials"
-	"github.com/aws/aws-sdk-go/aws/session"
-	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/aws"             //nolint:depguard
+	"github.com/aws/aws-sdk-go/aws/credentials" //nolint:depguard
+	"github.com/aws/aws-sdk-go/aws/session"     //nolint:depguard
+	"github.com/aws/aws-sdk-go/service/s3"      //nolint:depguard
```
Comment (rootulp) [inline pkg/trace/fileserver.go:22]: FYI we disabled depguard on `main`. Ref: https://github.com/celestiaorg/celestia-core/blob/0af1e151d474646f571de2eba387dc561d1dd60d/.golangci.yml#L5

We may consider doing the same on v0.34.x-celestia b/c IMO it's cumbersome to deal with. Alternatively, we can add `"github.com/aws/aws-sdk-go/` to https://github.com/celestiaorg/celestia-core/blob/05543c8ed486a9c09d0b3100a14d34c19e7c41ec/.golangci.yml#L56-L63

Comment (evan-forbes) [reply pkg/trace/fileserver.go]: good idea, fixed in https://github.com/celestiaorg/celestia-core/pull/1290/commits/7f4d81d330af738f7752b92bbd6606710cb32b77

**Review by cmwaters — APPROVED** [review]
I think we should be wary that we're using v0.34.x-celestia for both v1 and v2 which could be breaking in some ways i.e. with the config and could thus make it difficult to release patches if they were needed

File: config/config.go [hunk 2]
```diff
@@ -1188,24 +1188,24 @@ type InstrumentationConfig struct {
 	// Instrumentation namespace.
 	Namespace string `mapstructure:"namespace"`
 
-	// InfluxURL is the influxdb url.
-	InfluxURL string `mapstructure:"influx_url"`
+	// TracePushConfig is the relative path of the push config. This second
+	// config contains credentials for where and how often to.
```
Comment (rootulp) [inline config/config.go:1192]: ?

```suggestion
	// config contains credentials for where and how often to push.
```

File: config/config.go [hunk 3]
```diff
@@ -1188,24 +1188,24 @@ type InstrumentationConfig struct {
...
-	// InfluxToken is the influxdb token.
-	InfluxToken string `mapstructure:"influx_token"`
+	// TracePullAddress is the address that the trace server will listen on for
+	// pulling data.
+	TracePullAddress string `mapstructure:"trace_pull_address"`
 
-	// InfluxOrg is the influxdb organization.
-	InfluxOrg string `mapstructure:"influx_org"`
+	// TraceType is the type of tracer used. Options are "local" and "noop".
+	TraceType string `mapstructure:"trace_type"`
```
Comment (rootulp) [inline config/config.go:1200]: [question] why not continue supporting the existing config + pushing to InfluxDB and add the local tracer as a strictly new feature + new config? I think that could make this PR non-breaking.

File: config/config.go [hunk 4]
```diff
@@ -1188,24 +1188,24 @@ type InstrumentationConfig struct {
...
 
-	// InfluxBucket is the influxdb bucket.
-	InfluxBucket string `mapstructure:"influx_bucket"`
+	// TraceBufferSize is the number of traces to write in a single batch.
+	TraceBufferSize int `mapstructure:"trace_push_batch_size"`
```
Comment (rootulp) [inline config/config.go:1203]: [nit] why don't the field name and tag match?

```suggestion
	TraceBufferSize int `mapstructure:"trace_buffer_size"`
```

File: config/config.go [hunk 5]
```diff
@@ -1264,21 +1264,18 @@ func (cfg *InstrumentationConfig) ValidateBasic() error {
...
-	// if there is not InfluxURL configured, then we do not need to validate the rest
+	// if there is not TracePushConfig configured, then we do not need to validate the rest
 	// of the config because we are not connecting.
-	if cfg.InfluxURL == "" {
+	if cfg.TracePushConfig == "" {
 		return nil
 	}
-	if cfg.InfluxToken == "" {
+	if cfg.TracePullAddress == "" {
 		return fmt.Errorf("token is required")
```
Comment (rootulp) [inline config/config.go:1273]: [blocking] this error looks incredibly misleading given the field rename

```suggestion
		return fmt.Errorf("trace pull address is required")
```

On second thought, is this field optional or required? I suspect it's optional in which case the whole conditional can be deleted.

File: config/config.go [hunk 6]
```diff
@@ -1264,21 +1264,18 @@ func (cfg *InstrumentationConfig) ValidateBasic() error {
...
 	}
-	if cfg.InfluxOrg == "" {
+	if cfg.TraceType == "" {
 		return fmt.Errorf("org is required")
```
Comment (rootulp) [inline config/config.go:1276]: [blocking] again this error is incredibly misleading

```suggestion
	if cfg.TraceType == "" {
		return fmt.Errorf("TraceType is required")
```

File: config/toml.go [hunk 7]
```diff
@@ -547,25 +547,26 @@ max_open_connections = {{ .Instrumentation.MaxOpenConnections }}
 # Instrumentation namespace
 namespace = "{{ .Instrumentation.Namespace }}"
 
-# The URL of the influxdb instance to use for remote event
-# collection. If empty, remote event collection is disabled.
-influx_url = "{{ .Instrumentation.InfluxURL }}"
+# TracePushConfig is the relative path of the push config.
+# This second config contains credentials for where and how often to
```
Comment (rootulp) [inline config/toml.go:551]: idk what "second" means here? What was the first config?

File: config/config.go (hunk 5 above)
Comment (rootulp) [reply config/config.go:1273]: Copying code from elsewhere in this PR:

```
# The tracer pull address specifies which address will be used for pull based
# event collection. If empty, the pull based server will not be started.
```

seems optional so we shouldn't error here

File: pkg/trace/README.md [hunk 8]
```diff
@@ -1,136 +1,86 @@
...
 
-### Querying Data Using Python
-
-Python can be used to quickly search for and isolate specific patterns.
-
-```python
-from influxdb_client import InfluxDBClient
-from influxdb_client.client.write_api import SYNCHRONOUS
+Trace data will now be stored to the `.celestia-app/data/traces` directory, and
+save the file to the specified directory in the `table_name.jsonl` format.
```
Comment (rootulp) [inline pkg/trace/README.md:23]: grammar looks off

```suggestion
Trace data will now be stored to the `.celestia-app/data/traces` directory, and
files will be saved in the `table_name.jsonl` format.
```

File: pkg/trace/doc.go [hunk 9]
```diff
@@ -1,101 +1,2 @@
...
-
-To push events from e2e tests, we only need to specify the URL and the token via
-the cli.
-
-```bash
-cd test/e2e
-make && ./build/runner -f ./networks/ci.toml --influxdb-url=http://your-influx-ip:8086/ --influxdb-token="your-token"
-```
-*/
+/**/
```
Comment (rootulp) [inline pkg/trace/doc.go:1]:
```suggestion
```

**Review by rootulp — COMMENTED** [review]
[question] why do we want to backport a config breaking change to v0.34.x? How do we intend on releasing this change? Will it be v2.0.0-tm-v0.34.29? 

If that's the case, what's our current branch management strategy in this repo because I think this backport conflicts with my understanding:

> v0.34.x-celestia was originally based on tendermint's v0.34.x release branch but now it receives patches from the CometBFT v0.34.x release branch. This branch also contains Celestia-specific changes. Future v1.x.0-tm-v0.34.x releases of this repo will be based on this branch.

Ref: https://github.com/celestiaorg/celestia-core?tab=readme-ov-file#branches

**Comment by evan-forbes** [comment]
> I would still like to understand the release strategy for this breaking change.

this is a good point, and we should document this. in the past, we've briefly discussed using a migration script. In a similar vein to a migration script, if we know what the error is, then we can automatically fix the config and log it should that error occur.

File: pkg/trace/README.md (hunk 8 above)
Comment (evan-forbes) [reply pkg/trace/README.md:23]: fixing upstream and moving down

File: config/toml.go (hunk 7 above)
Comment (evan-forbes) [reply config/toml.go:551]: the second config is a json that contains creds to accessing an s3 bucket

File: config/config.go (hunk 6 above)
Comment (evan-forbes) [reply config/config.go:1276]: fixing upstream and moving down

File: config/config.go (hunk 4 above)
Comment (evan-forbes) [reply config/config.go:1203]: just forgot to change

File: config/config.go (hunk 3 above)
Comment (evan-forbes) [reply config/config.go:1200]: we have to have a fix for #1141 , which is already breaking, but also keeping the old tracer was pretty burdensome as we had to have a ton of consts and enforce the schema in a really weird way. Here, we're just using a struct, but if we keep the influx, we have to keep two separate mechanisms for enforcing the schema

**Comment by rootulp** [comment]
During a 1:1 I think we decided to make the config change strictly additive and non-breaking so that we don't have to write a config migrator. If that's the case, do we still want this PR reviewed as-is?

**Review by staheri14 — APPROVED** [review]

**Comment by evan-forbes** [comment]
going to merge this, as I just tested and this is no longer breaking the config now that we've renamed everything. its still a super breaking change in that we've technically replaced an entire feature with something that is completely different, but we don't have to complete #1299

**Comment by rootulp** [comment]
@evan-forbes why did we merge this? As far as I can tell, it's a breaking change and it targeted a release branch. How do we intend on cutting the next release from v0.34.x-celestia?

**Pull request merged**